README.md
Congratulations!  You are reading the README file for a software
package you downloaded!  You are now officially an *advanced* software
installer.

This is the README file for OOF2, describing how to build and install
it with the Python distutils utility.

This README file is for OOF2 version 2.1.12 or later.

Disclaimer:
----------

This software provided is provided by NIST as a public service. You
may use, copy and distribute copies of the software in any medium,
provided that you keep intact this entire notice. You may improve,
modify and create derivative works of the software or any portion of
the software, and you may copy and distribute such modifications or
works. Modified works should carry a notice stating that you changed
the software and should note the date and nature of any such
change. Please explicitly acknowledge the National Institute of
Standards and Technology as the source of the software.  To facilitate
maintenance we ask that before distributing modified versions of this
software, you first contact the authors at oof_manager@nist.gov.

The software is expressly provided "AS IS". NIST MAKES NO WARRANTY OF
ANY KIND, EXPRESS, IMPLIED, IN FACT OR ARISING BY OPERATION OF LAW,
INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTY OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT
AND DATA ACCURACY. NIST NEITHER REPRESENTS NOR WARRANTS THAT THE
OPERATION OF THE SOFTWARE WILL BE UNINTERRUPTED OR ERROR-FREE, OR THAT
ANY DEFECTS WILL BE CORRECTED. NIST DOES NOT WARRANT OR MAKE ANY
REPRESENTATIONS REGARDING THE USE OF THE SOFTWARE OR THE RESULTS
THEREOF, INCLUDING BUT NOT LIMITED TO THE CORRECTNESS, ACCURACY,
RELIABILITY, OR USEFULNESS OF THE SOFTWARE.

You are solely responsible for determining the appropriateness of
using and distributing the software and you assume all risks
associated with its use, including but not limited to the risks and
costs of program errors, compliance with applicable laws, damage to or
loss of data, programs or equipment, and the unavailability or
interruption of operation. This software is not intended to be used in
any situation where a failure could cause risk of injury or damage to
property. The software was developed by NIST employees. NIST employee
contributions are not subject to copyright protection within the
United States.

Installation:
------------

Installation is similar to other Python libraries.  If your system is
well-set-up, and has the required libraries, there should be no
difficulties.

The executive summary of steps is:

> tar -xzf oof-<version>.tar.gz
> cd oof-<version>
> python setup.py build
> python setup.py install

but please read the rest of this file before proceeding.

If something goes wrong, your system adminstrator may be able to help
you, or you can contact the oof developers at oof_manager@nist.gov.
It's diagnostically useful to include all the output from setup.py.

OOF2 has been built and tested on Linux (Debian, CentOS, Ubuntu, and,
OpenSUSE), and Macintosh OS X.

Details:
-------

0. What You Should Already Have:

   A computer running a variant of the Unix operating system,
   including Linux and Macintosh OS X, and an X11 server.  OOF2
   currently does *not* run on Microsoft Windows.
   
   The following external programs and libraries must be present
   before you can run OOF2.  To compile OOF2 from sources, you will
   also require the header files ("includes") associated with these
   programs and libraries.  These are usually available as part of a
   "development" version of the library software.

    Python (2.6 or 2.7)      http://www.python.org 
    Magick++                 http://www.imagemagick.org/www/Magick++/index.html
    gtk+-2.0 (2.6 or later. Not 3.x)  http://www.gtk.org/download/   
    libgnomecanvas2   http://directory.fsf.org/graphics/misc/libgnomecanvas.html
    pygtk2 (2.6 or later)    http://www.pygtk.org

   Macintosh OS X users will need to install an X11 server:
   	     	  	     http://xquartz.macosforge.org/

   You should also have the ability to run lapack and the "blas" basic
   linear algebra subroutines.  On Macintosh OS X, they are built in
   to the Accelerate framework in the OS, and no special libraries are
   required.  On Linux and commercial Unix systems, they may have to
   be installed, and you may require headers (sometimes provided as
   part of a "-dev" package).

   Detailed instructions for installing the OOF2 prerequisites on a
   number of different operating systems can be found at
   http://www.ctcms.nist.gov/oof/oof2/prerequisites.html.

1. What You Got From Us:
   
   The file "oof2-<version>.tar.gz".


Procedure:
---------

   (Macintosh OS X users can install OOF2 from either a Terminal or
   xterm window, or the equivalent.)

0. Unpack

   Unpack the .tar.gz file.  The usual way is to run "tar -xzf" on the
   file you want to unpack.  This will create a subdirectory named
   "oof2-<version>" in the directory where you run tar.

1. Build the OOF2 libraries and Python extension modules

   Switch to the newly-created directory, and run

     % python setup.py build

   The build_ext command will create a "build" subdirectory in the top
   OOF2 directory.  Within "build" it will create a subdirectory with
   a system-dependent name.

1.1 Getting more control over the build

   You can ignore this section unless something went wrong when
   building OOF2 in step 1.  setup.py tries to be intelligent about
   choosing options, but it's not perfect.

   The distutils "build" command actually runs a bunch of separate
   subcommands, each of which has its own options.  The relevant
   subcommands are "build_shlib", "build_ext", "build_scripts", and
   "build_py".  "build_shlib" builds the shared libraries,
   liboof2common.so, etc, that contain most of the low-level OOF2
   machinery. "build_ext" builds the OOF2 Python extension modules
   that provide the interface betweeen C++ and Python.  "build_py"
   copies the Python files from the source directory to the build
   directory, and "build_scripts" copies the start-up script into the
   build directory and makes it executable.  OOF2 installers will
   probably only have to worry about "build_shlib" and "build_ext".

   The four commands must be run in order: build_shlib must precede
   build_ext, and build_ext must precede build_py.

   Each command can be run separately, for example

     % python setup.py build_ext

   or in combination

     % python setup.py build_shlib build_ext

   and options can be provided to each one
  
    % python setup.py build_shlib --debug build_ext --include_dirs=/sw/include

   You can see the full set of options by running

     % python setup.py --help <command name>

   Here are the options most likely to be useful:

   For "build_shlib" or "build":

   --library-dirs    Specify a non-standard location for libraries.
                     Multiple directories should be separated by
                     colons, like this:
                       --library-dirs=/strange/spot:/out/of/theway

   --libraries       Specify libraries to load. Due to a bug in
                     distutils, it's only possible to specify a single
                     library. For example
                        --libraries=abc
		     will load libabc.so.  If you need to load more
                     than one library in this way, please contact us.
      
   --blas-libraries  Specify libraries to use for blas and lapack.
                     Multiple library names should be separated by
                     spaces, like this:
                        --blas-libraries="myblas mylapack"

   --blas-link-args  Specify additional link arguments required by
                     blas and lapack, for example:
		        --blas-link-args="-faltivec -framework vecLib"

   For "build_ext" or "build":

   --skip-swig       Don't run swig.  The oof2 distribution includes
                     the swig output files, so it's not necessary to
                     run swig unless you've changed the swig input
                     files.  oof2 also requires an old version of swig
                     (1.1 build 883), and if you accidentally run a
                     newer version you will corrupt the distributed
                     files.  If that happens, unpack the tar file
                     again (to get a clean copy) and make sure to
                     specify --skip-swig when you run build or
                     build_ext.  See the note in item 2, below. You
		     *must* run the build and install steps
		     simultaneously if you use --skip-swig.

   --with-swig       Specify the name of the swig executable.  Swig is
                     a tool that generates the code that allows
                     communication between python and C++.  OOF2 now
                     ships with a modified copy of swig, because we
                     rely on an old version (swig 1.1 build 883, to be
                     precise).  If you have another version of swig
                     that you'd like to use, you can name it here,
                     like this:
                        pyton setup.py build --with-swig=myswig
                     Be advised that it probably won't work with the oof2
                     source code. 

   --swig, --swig-cpp, --swig-opts
                     Do NOT use these options.  They will be ignored.
                     They're leftover from the raw distutils swigging
                     mechanism, which we don't use.

   The following arguments can appear anywhere after "setup.py" in the
   command line, and apply to both the build and install steps.  Tf
   you run the build and install steps separately, you must provide
   these arguments in *both* steps if you provide them in one.

   --disable-gui     Don't include any components of the graphical user
                     interface.  When this option is used, it's not
                     necessary to have the gtk, pygtk, or
                     libgnomecanvas libraries installed.

2. Install

   To install OOF2, run

   % python setup.py install

   This will install OOF2 in the standard location for Python
   extensions on your system.  This is good, because then you won't
   have to do anything special to get OOF2 to run.  It's also bad,
   because unless you are the system administrator, you probably don't
   have permission to install anything in that directory.  You have
   two options:

    a) Get a system administrator to run the installation step.

    b) Tell distutils to install oof2 in a different place, like this:

          % python setup.py install --prefix=<prefix>

    where <prefix> is a directory that you can write to.  The default
    value of <prefix> is usually /usr/local.  On OS X it may be
    something like /Library/Frameworks/Python.framework/Versions/2.7
    if you're using the system Python, or /sw or /opt/local if you're
    using fink or macports.

    The installation procedure will create an executable script called
    "oof2" in <prefix>/bin, a bunch of shared libraries called
    "liboof2*.so" or "liboof2*.dylib" in <prefix>/lib, a directory
    called "oof2" in <prefix>/lib/python2.x/site-packages (where 2.x
    is your python version number), and some example files in
    <prefix>/share/oof2/examples. 

    (It's possible to use --home=<home> instead of --prefix when
    installing oof2.  The only difference is that --home will put the
    python libraries in <home>/lib/python instead of
    <prefix>/lib/python2.x/site-packages.)

2.1. Set environment variables

    If <prefix>/bin is not in your Unix command path, you'll need to
    add it to the PATH environment variable, or create a symbolic link
    from a directory that is in your path (or start OOF2 the hard way
    by by typing <prefix>/bin/oof2).  (Typing "echo $path" will print
    the current value of your path.  The method for setting
    environment variables depends on which Unix shell you're using.)

    If <prefix>/lib is not in the list of directories that the dynamic
    linker searches for libraries, you'll have to add it by setting
    the LD_LIBRARY_PATH environment variable.  This should *not* be
    necessary on Macintosh OS X.

    If <prefix>/lib/python2.x/site-packages is not in your Python
    path, you'll have to add it to the PYTHONPATH environment
    variable.   (Running the command
       % python -c "import sys; print sys.path"
    will print your Python path.)


Running OOF2:
------------

At this point, you should have an executable file named "oof2" in a
bin directory in your execution path.  You can now simply type "oof2"
at your shell prompt, and OOF2 will start up.

OOF also has many options, and you can get a summary of them by typing
"oof2 --help".

By default, OOF runs in graphics mode, opening a couple of windows to
get you started.  If you don't want this, you can use the "--text"
option to run it in command-line mode.

Be sure to read the OOF manual, at
  http://www.ctcms.nist.gov/~langer/oof2man/
and to go through the tutorials provided in the OOF2 help menu.


Contact Us:
----------

If you encounter bugs in the program, please send e-mail to
oof_bugs@nist.gov.  Tell us what version of OOF2 you're using, what
operating system you're using, and *exactly* what you did to encounter
the error.  It is helpful to include an OOF2 script (which you can
save with the "File/Save/Python Log" menu item) and a copy of any
input files (images, oof data files, etc) required to run the script.
It is extremely difficult for us to fix a bug if we can't reproduce it
here. 

Other communications, including requests for help and suggestions for
new features, can be sent to oof_manager@nist.gov.


SRC/changelog.txt
OOF2 2.1.12

* Fixed setup.py so that it understands sys.platform==linux3.

* Fixed a bug in the Range output.

* A modified version of swig is now included in the OOF2 distribution,
  and the setup.py script now builds and uses it.  The --skip-swig
  option is no longer required (or allowed).

* Fixed a bug in the handling of invalid inputs to
  LogDefiniteProgress::setFraction().  This may have never occurred.

* Fixed a bug in pixel group comparison that was effectively making
  all unmeshable groups meshable.

OOF2 2.1.11

* Fixed a bug introduced in 2.1.10 by being too aggressive about
  cleaning up unused code that wasn't actually unused.

OOF2 2.1.10

* Fixed a possible divide-by-zero error that could arise from an
  unlikely and perverse choice of end points for a cross section.

* Modifications for compatibility with the clang++ compiler on OS X
  10.9.

* Fixed a theoretical but unobserved bug that could occur when two
  Skeleton edge boundaries wrapped around a single element in both
  directions. 

OOF2 2.1.9

* The Save button on the Microstructure page now allows the file
  format to be selected.

* Fixed a bug in abaqus output of skeletons with materials.

OOF2 2.1.8

* Reduced the memory footprint of dense-matrix multiplications.

* Added a "Pause" command to the GUI Logging menu.

* Fixed a bug in copying Neumann boundary conditions.

* Hacked the build scripts to accomodate the Gentoo "prefix" package
  manager's hacked version of distutils.

* Fixed the pole figure test in the TEST directory.  Twice.

OOF2 2.1.7

* Added a build time check for GraphicsMagick, which is meant to be
  compatible with ImageMagick, but isn't.

* Worked around a problem in which the setup script couldn't find
  pygtk on systems using MacPorts with Python 2.6 or later.

* Fixed bugs and misconceptions in the generation of pole figures.

OOF2 2.1.6

* Fixed leaky memory in image modification.

* Fixed a bug in reporting malformed materials.

* Changes required for the clang compiler in OS X 10.7.

* Added a "Generic" orientation map reader that ought to be able to
  read most EBSD files if they contain ascii lines with xy and
  orientation data in them.  It can flip the input data in the x and y
  directions and apply a constant offset to the orientations.

* Added the option to flip data and offset the orientations to the TSL
  and TSL2 orientation map readers.

* Added a Pole Figure generation command to the Orientation Map menu.

* Fixed bug in setting default display parameters for pixel
  selections.

* Fixed the skeleton boundary display, which was using the arrow width
  instead of the line width for all but the first segment of the
  displayed edge boundaries.

OOF2 2.1.5

* Fixed the ascii output for rank 2 tensors so that it doesn't lose
  precision for very small values.

* Fixed a Tetragonal elasticity bug.  The modulus was being saved
  incorrectly in binary data files.

OOF2 2.1.4

* Fixed another bug that caused crashes when using orientation maps
  with particular Mesh geometries.  Round-off error was causing the
  OrientationMapProperty to try to evaluate an orientation at a pixel
  outside of the map.

* Fixed a crash triggered by using the 9-noded quadrilateral element.
  You shouldn't use this element anyway, since the 8-noded quad is
  cheaper and equally accurate.

* Changed the default values of the Hexagonal and Tetragonal elastic
  moduli so that they're not singular.  The default values are now
  effectively isotropic.

* Fixed a bug that could cause the program to hang when computing
  contour plots.

* Fixed some memory leaks in contour plots.

OOF2 2.1.3

* Fixed a bug that caused crashes when using orientation maps,
  plotting stresses, and rebuilding meshes.

OOF2 2.1.2

* The interpretation of the heat source property has been changed, to
  make it sane.  A positive heat source adds heat to the system. 

* Fixed bugs related to time dependent boundary conditions in
  quasistatic problems.

* Added a mechanism for setting the initial values for floating
  boundary conditions in time dependent problems.

OOF2 2.1.1

* Scheduled Output data files are no longer automatically rewound,
  ever.  If repeating a calculation, the files need to be rewound
  manually if the results of the first calculation should be
  overwritten. 

* Typing in the time entry box in a graphics window is friendlier.  It
  was moving the cursor to the left after every keystroke.

* New images created from orientation maps are now displayed in the
  graphics window if there are no other images present.

* Added a new orientation map reader, "TSL2", which can read some more
  modern TSL .ang files.  There is no good way to determine which
  reader to use.

* TSL orientation map data can contain negative xy coordinates.
  Points no longer need to be provided in a particular order.

* The "Existing Stream" option for specifying an destination for output
  is no longer available.  Use "Output Stream" instead.  Multiple
  Output Streams can point to the same file.  If they have different
  modes, the most recently selected mode will be used.

* Fixed crashing bugs in abaqus output for skeletons and in creating
  skeleton element, node, and segment groups from pixel groups.

* Changed the Schedule_Solution and Scrub_Solution commands in the
  OOF.Subproblem menu to Enable_Solution and Disable_Solution.

* Moved Copy_All_Solvers from the Subproblem menu to the Mesh menu.

* Fixed a setup problem with Python 2.7.

* Fixed scheduled outputs so that multiple outputs sent to a single
  file are always written in the same order.

OOF2 2.1.0

* Time dependence!

* On OS X, OOF2 now links to the Accelerate framework instead of the
  deprecated vecLib framework.

* Fixed a long standing bug that caused an error when invoking a
  togglable menu (eg, ActivityViewer/Settings/AutoDismiss) item in a
  re-opened window.

* The Mesh Info toolbox will now give info for empty elements

oof2setuputils.py
# -*- python -*-
# $RCSfile: oof2setuputils.py,v $
# $Revision: 1.19 $
# $Author: langer $
# $Date: 2014/07/04 20:16:39 $


# This software was produced by NIST, an agency of the U.S. government,
# and by statute is not subject to copyright in the United States.
# Recipients of this software assume all responsibilities associated
# with its operation, modification and maintenance. However, to
# facilitate maintenance we ask that before distributing modified
# versions of this software, you first contact the authors at
# oof_manager@nist.gov.

# This file contains routines used by the main OOF2 setup.py script.
# These routines are also used by oof2extutils.py, which is used by
# the authors of external OOF2 extensions.

import os, sys, string, stat

# Utilities for running xxxxx-config programs that get the flags and
# libs required to link with external libraries.  "clib" is a CLibInfo
# object. "cmd" is a command to run that returns the required data,
# such as "Magick++-config --cflags".

def add_third_party_includes(cmd, clib):
    f = os.popen(cmd, 'r')
    for line in f.readlines():
        for flag in line.split():
            if flag[:2] == '-I':
                clib.includeDirs.append(flag[2:])
            else:
                clib.extra_compile_args.append(flag)

def add_third_party_libs(cmd, clib):
    f = os.popen(cmd, 'r')
    for line in f.readlines():
        for flag in line.split():
            if flag[:2] == '-l':
                clib.externalLibs.append(flag[2:])
            elif flag[:2] == '-L':
                clib.externalLibDirs.append(flag[2:])
            else:
                clib.extra_link_args.append(flag)

# Check for packages that use pkg-config for their options.  Include
# their compiler and linker flags if they're found, and complain if
# they're not.

def pkg_check(package, version, clib=None):
    if check_exec('pkg-config'):
        if os.system("pkg-config --atleast-version=%s %s" % (version, package)):
            print "Can't find %s! Version %s or later required" % (package,
                                                                   version)
            sys.exit()
        if clib:
            add_third_party_libs("pkg-config --libs %s" % package, clib)
            add_third_party_includes("pkg-config --cflags %s" % package, clib)
    else:
        print "Can't find pkg-config!"
        sys.exit()
    


# Check that an executable file exists in the environment's PATH.

pathdirs = os.getenv("PATH").split(":")

def check_exec(xfile):          # is xfile an executable file in the path?
    for dir in pathdirs:
        try:
            if os.path.exists(dir):
                files = os.listdir(dir)
                if xfile in files:
                    fullname = os.path.join(dir, xfile)
                    if os.access(fullname, os.X_OK):
#                         print >> sys.stderr, "Found", xfile, "as", fullname
                        return 1
        except: # Ignore permission errors
            pass
    if sys.platform == 'cygwin' and not xfile.endswith('.exe'):
        return check_exec(xfile + '.exe')
    print "Warning!", xfile, "not found!"

# See if a given file exists in a list of directories.  Return the
# directory containing the file, or None if the file can't be found.

def find_file(filename, dirs):
    for dir in dirs:
        if os.path.exists(os.path.join(dir, filename)):
            return dir

# Add dirs to the given path environment variable.

def extend_path(envpath, *dirs):
    try:
        pathstring = os.environ[envpath]
    except KeyError:
        path = []
    else:
        path = pathstring.split(':')
    # prepend directories to the path
    path[0:0] = [d for d in dirs if os.path.exists(d)]
    if path:
        os.environ[envpath] = ':'.join(path)
    
###################

SWIG = 'swig'                           # must be swig 1.1 build 883
SWIGARGS = ["-shadow", "-dnone", "-python", "-c++", "-c"]

def run_swig(srcdir, swigfile, destdir, cext="_.C", include_dirs=[],
             dry_run=False, force=False, extra_args=[],
             with_swig=None, DIM_3=0):

    # srcdir is a directory.  swigfile is the name of the input swig
    # file, specfied *relative* to srcdir.  The output files will go
    # into destdir.  If swigfile has subdirectories as part of its
    # name, then those subdirectories will be created in destdir.  For
    # example, if srcdir is "A/B", swigfile is "C/swig.swg", and
    # destdir is "E/F", then the file "A/B/C/swig.swg" will be swigged
    # to create "E/F/C/swig_.C" and "E/F/C/swig.py".

    # cext is the extension to use on the output C++ file.

    swig = with_swig or SWIG

    srcdir = os.path.abspath(srcdir)
    destdir = os.path.abspath(destdir)
    relpath = os.path.split(swigfile)[0]
    infile = os.path.join(srcdir, swigfile)
    indir = os.path.join(srcdir, relpath)
    outdir = os.path.join(destdir, relpath)

    swigfilename = os.path.split(swigfile)[1]
    basename = os.path.splitext(swigfilename)[0]
    cfile = os.path.join(destdir, relpath, basename + cext)
    pyfile = os.path.join(destdir, relpath, basename + ".py")

    infiletime = os.stat(infile)[stat.ST_MTIME]
    uptodate = os.path.exists(cfile) and os.path.exists(pyfile) and \
               (os.stat(cfile)[stat.ST_MTIME] > infiletime) and \
               (os.stat(pyfile)[stat.ST_MTIME] > infiletime)

    if DIM_3 and "-DDIM_3" not in extra_args:
        extra_args.append("-DDIM_3")
    
    if force or not uptodate:

        # Make sure that outdir exists.
        if not os.path.exists(outdir):
            os.makedirs(outdir)
        # Make sure that all directories from destdir to outdir
        #  contain __init__.py.
        destdepth = len(destdir.split(os.sep))
        outsplit = outdir.split(os.sep)[destdepth:]
        addInitPy(destdir)
        odir = destdir
        for subdir in outsplit:
            odir = os.path.join(odir, subdir)
            addInitPy(odir)

        incls = [os.path.abspath(dir) for dir in include_dirs]
        if srcdir not in incls:
            incls.append(srcdir)
        if indir not in incls:
            incls.append(indir)
        
        # The following command cds to the source directory and runs
        # swig.  It uses absolute path names for all of the
        # directories so that we don't have to worry about the effect
        # of the cd.  The cd is necessary because the swig
        # 'pragma(python) include' lines don't seem to respect the -I
        # flags.
        
        cmd = 'cd %(indir)s; %(swig)s %(swigargs)s %(includes)s -module %(module)s -o %(cfile)s %(infile)s' \
              %  {'indir' : indir,
                  'swig': swig,
                  'swigargs' : string.join(SWIGARGS+extra_args, " "),
                  'includes' : string.join(['-I'+file for file in incls], ' '),
                  'module' : basename,
                  'cfile' : cfile,
                  'infile' : infile}
        print cmd
        if not dry_run:
            os.system(cmd)

    return dict(basename=basename,
                indir=indir,
                outdir=outdir,
                cfile=cfile
                )


def addInitPy(subdir):
    initpy = os.path.join(subdir, '__init__.py')
    if not os.path.exists(initpy):
        print >> sys.stderr, "Creating", initpy
        initfile = open(initpy, 'w')
        print >> initfile, "# This file no verb"
        initfile.close()
